### PR TITLE
issue #1288 User home folder do not need to read all of its children.

### PR DESCRIPTION
**Pull request by lmcbout**
Signed-off-by: Jacques Bouthillier <jacques.bouthillier@ericsson.com>

### CODE REVIEW

**Comment by hexa00** [comment]
As discussed the goal is that:

0 is used as default
1 is used on a needed basis

Seems like you're missing commits for:
   - Analysis to check if 1 is necessary on the calls that are using 1
   - Default to 0 

But it's a start.

**Comment by marcdumais-work** [comment]
I find that the PR title and commit message do not reflect well the content of the PR. 

For the commit message, you should not refer to the issue. instead you can add "fixes #1298" in the body of the commit message.

**Comment by kittaakos** [comment]
I will check the changes, but first I would like to reproduce the defect.

**Comment by hexa00** [comment]
Alternatively I think it would be nice to have children's computed lazily. 

Such that a call to .children reads them.

This would avoid having the option in the call.

File: packages/filesystem/src/common/filesystem.ts [hunk 1]
```diff
@@ -17,10 +17,11 @@ export interface FileSystem extends JsonRpcServer<FileSystemClient> {
      * Returns the filestat for the given uri.
      *
      * If the uri points to a folder it will contain one level of unresolved children.
+     * The depth define the number of level for unresolved children.
      *
      * Reject if a file for the given uri does not exist.
      */
-    getFileStat(uri: string): Promise<FileStat>;
+    getFileStat(uri: string, depth: number): Promise<FileStat>;
```
Comment (kittaakos) [inline packages/filesystem/src/common/filesystem.ts]: Without assuming anything about the implementation, what should happen if I pass in `2` or `3`?

Comment (kittaakos) [reply packages/filesystem/src/common/filesystem.ts]: Can we have an optional (or default) value maybe?

**Comment by kittaakos** [comment]
Let's put aside the fact that we could do lazy children loading... I must be blind, but what is the actual problem? Why do we need this PR?

**Comment by kittaakos** [comment]
Even if we need this fix, but I do not see yet the problem, you can do just this instead, right?
```typescript
    async getCurrentUserHome(): Promise<FileStat> {
        const uri = FileUri.create(os.homedir());
        const stat = await this.doGetStat(uri, 0);
        if (!stat) {
            throw new Error(`Cannot find file under the given URI. URI: ${uri}.`);
        }
        return stat;
    }
```

**Comment by kittaakos** [comment]
Also, this test would fail:
```typescript
  it("should have consistent children", async () => {
    const actual = ((await createFileSystem().getCurrentUserHome()).children || []).length;
    const expected = fs.readdirSync(os.homedir()).length;
    expect(expected).to.be.equal(actual);
  });
```

You cannot just break the API contract for the home folder because it is slow with a huge number of files.

**Comment by kittaakos** [comment]
> Alternatively I think it would be nice to have children's computed lazily.
>
> Such that a call to .children reads them.
> 
> This would avoid having the option in the call.

Just keep in mind, this should go through the wire as a JSON, right? (That is why we do not send URI objects either.) So we would need to serialize the function, which is the `.toString()`. AFAIK, most browsers support it, but we could easily hit a native call (`fs`?) which is `[native code]`. And I am not sure if closures are supported at all and how it works with `inversify`. But please correct me if I am wrong or there is a feasible approach.

**Comment by hexa00** [comment]
@kittaakos 

>Let's put aside the fact that we could do lazy children loading... I must be blind, but what is the actual problem? 

The problem is that doing getFileStat on a directory can be slow because of the number of children and that it is not necessary in some cases, the home folder is only one of them. @lmcbout will document more or these.

The idea was to have the default getFileStat with depth 0 and require more depth as needed or redesign the API a bit for things to be lazy.

I do agree that it would make more sense as `getFileStat(resolveChildren?: boolean)` instead of the depth.

The idea was that getFileStat gives you information about the file/directory you pass in argument and nothing else by default, and if you want more you need to specify it. 

Otherwise it's too easy to miss-use it you 're thinking you're only affected one file or directory but it reads all the children which is non intuitive. No filesystem API work like that.

For the lazy part yes it's just an idea to investigate I'm not sure it's best with json-rpc to have it as a accessor, it could be a distinct method however like getFileChilds() or something like that.

**Comment by kittaakos** [comment]
I agree, lazy loading children would be nice, but both the task and the PR focused on the home folder issue. Please open a follow-up task for that or clarify the requirements at #1288. Then we can figure out how to tackle the problem. With the current changes, I cannot agree.

**Comment by hexa00** [comment]
@kittaakos yes @lmcbout will change the issue etc based on the findings that other things are affected.

**Comment by lmcbout** [comment]
As suggested by Antoine, I put a default value to be "0" .

I discover something else. When a file is saved on the client side, node-filesystem.doGetStat() is executed for the file and the function is also executed for the whole workspace, so all files try to get the FileStat again for no reason. If I save a file, we can get the FileStat for the file, but certainly no need to get the FileStat for all files in the workspace. When the auto-save is on, looking for the FileStat for the workspace is executed a lot. To see how often it is used, just put the following line in node-filesystem.doGetStat at line 280.
console.log("doGetStat() depth: " + depth + "  for: " + uri.toString());

**Comment by svenefftinge** [comment]
A default of 0 doesn't make sense if all but the home folder references explicitly pass `1`.
Also, do you have numbers for how often and when they are fetched unnecessarily? Isn't the home folder only fetched once atm?

**Comment by marcdumais-work** [comment]
@svenefftinge looks like something similar is happening at every save. see https://github.com/theia-ide/theia/issues/1300#issuecomment-366388612

Also when I open a WS, I see this, in addition to what happens in the home folder: https://github.com/theia-ide/theia/issues/1288#issuecomment-366391099

File: packages/filesystem/src/common/filesystem.ts [hunk 2]
```diff
@@ -17,10 +17,11 @@ export interface FileSystem extends JsonRpcServer<FileSystemClient> {
      * Returns the filestat for the given uri.
      *
      * If the uri points to a folder it will contain one level of unresolved children.
+     * The depth define the number of level for unresolved children. Default value is 0
      *
      * Reject if a file for the given uri does not exist.
      */
-    getFileStat(uri: string): Promise<FileStat>;
+    getFileStat(uri: string, depth?: number | 0): Promise<FileStat>;
```
Comment (akosyakov) [inline packages/filesystem/src/common/filesystem.ts]: - I would prefer to use the options pattern for optional arguments.
- type number includes type 0, `| 0` is redundant 
- As Sven mentioned 1 is the better default.

**Comment by kittaakos** [comment]
Can we close this in favor of #1302?

**Pull request closed without merging**